ZOS deployments test suite

samaradel opened the pull request:
### Description

Create a test suite for ZOS deployments

### Changes

Create a ZOS deployment test suite for:
- SignatureRequest
- Signature
- SignatureRequirement
- Deployment
- Deployment signing
- KeypairType enum

### Related Issues

- https://github.com/threefoldtech/tfgrid-sdk-ts/issues/2449

### Tested Scenarios

- Run gridclient > tests > zos > deployment file

### Documentation PR

For UI changes, Please provide the Documentation PR on [info_grid](https://github.com/threefoldtech/info_grid)

### To consider

#### Preliminary Checks:

- [x] Preliminary Checks
  - Does it completely address the issue linked?
  - What about edge cases?
  - Does it meet the specified acceptance criteria?
  - Are there any unintended side effects?
  - Does the PR adhere to the team's coding conventions, style guides, and best practices?
  - Does it integrate well with existing features?
  - Does it impact the overall performance of the application?
  - Are there any bottlenecks or slowdowns?
  - Has it been optimized for efficiency?
  - Has it been adequately tested with unit, integration, and end-to-end tests?
  - Are there any known defects or issues?
  - Is the code well-documented?
  - Are changes to documentation reflected in the code?

#### UI Checks:

- [x] UI Checks
  - If a UI design is provided/ does it follow it?
  - Does every button work?
  - Is the data displayed logical? Is it what you expected?
  - Does every validation work?
  - Does this interface feel intuitive?
  - What about slow network? Offline?
  - What if the gridproxy/graphql/chain is failing?
  - What would a first time user have a hard time navigating here?

#### Code Quality Checks:

- [x] Code Quality Checks
  - Code formatted/linted? Are there unused imports? .. etc
  - Is there redundant/repeated code?
  - Are there conditionals that are always true or always false?
  - Can we write this more concisely?
  - Can we reuse this code? If yes, where?
  - Will the changes be easy to maintain and update in the future?
  - Can this code become too complex to understand for other devs?
  - Can this code cause future integration problems?

### Testing Checklist

- [x] Does it Meet the specified acceptance criteria.
- [x] Test if changes can affect any other functionality.
- [ ] Tested with unit, integration, and end-to-end tests.
- [ ] Tested the un-happy path and rollback scenarios.
- [ ] Documentation updated to meet the latest changes.
- [ ] Check automated tests working successfully with the changes.
- [ ] Can be covered by automated tests.

### General Checklist

- [x] Tests included
- [x] Build pass
- [ ] Documentation
- [x] Code format and docstring
- [ ] Screenshots/Video attached (needed for UI changes)

Commit:
Create zos deployment test suite for:
- SignatureRequest
- Signature
- SignatureRequirement
- Deployment
- Deployment signing
- KeypairType enum

0oM4R reviewed (commented):
Great job, please consider adding edge cases like 
- Can we add a test case to the fromHex method, like passing an invalid hex
- `deployment.sign` , can we add a test to verify the result signature, and also edge cases, like in a valid signature 
- Verify the challenge format with no workloads

Commit:
- Add test case for fromHex method
- Add tests for signature verification
- Verify the challenge format with no workloads

amiraabouhadid commented on packages/grid_client/tests/zos/deployments.test.ts:
```diff
@@ -0,0 +1,773 @@
+      expect(deployment.signature_requirement.signatures[0].twin_id).toBe(0);
+    });
+
+    it("should handle signing with negative twin_id", async () => {
```
can we add one where twin_id is null/undefined?

Commit:
Add test cases for null, undefined twin id values

amiraabouhadid reviewed (commented):
all tests pass
![Screenshot 2025-07-01 at 16 57 52](https://github.com/user-attachments/assets/1125385b-8d7b-4f88-bba2-3aa85a0de7db)

amiraabouhadid commented on packages/grid_client/tests/zos/deployments.test.ts:
```diff
@@ -0,0 +1,827 @@
+
+      const signatureBytes = new Uint8Array(signature.length / 2);
+      for (let i = 0; i < signature.length / 2; i++) {
+        signatureBytes[i] = parseInt(signature.slice(2 * i, 2 * i + 2), 16);
+      }
+
+      const messageBytes = new Uint8Array(message.length / 2);
+      for (let i = 0; i < message.length / 2; i++) {
+        messageBytes[i] = parseInt(message.slice(2 * i, 2 * i + 2), 16);
+      }
```
this should probably be extracted in a util file as it may be reused 
also 
```suggestion
        const signatureBytes = new Uint8Array(signature.length / 2);
for (let i = 0; i < signature.length; i += 2) {
  signatureBytes[i / 2] = parseInt(signature.slice(i, i + 2), 16);
}

const messageBytes = new Uint8Array(message.length / 2);
for (let i = 0; i < message.length; i += 2) {
  messageBytes[i / 2] = parseInt(message.slice(i, i + 2), 16);
}
```

amiraabouhadid commented on packages/grid_client/tests/zos/deployments.test.ts:
```diff
@@ -0,0 +1,827 @@
+  NETWORK_NAME: "test_network",
+  NETWORK_SEED: "test_seed",
+
+  MNEMONIC: "abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about",
```
can we just import the mnemonics used in config.json instead of using these static dummy mnemonic?

amiraabouhadid commented on packages/grid_client/tests/zos/deployments.test.ts:
```diff
@@ -0,0 +1,827 @@
+  WORKLOAD_DESCRIPTION: "workload_description",
+
+  CPU: 1,
+  MEMORY: 1024 * 1024 * 1024, // 1GB in bytes
```
```suggestion
  MEMORY: 1024**3,  // 1GB in bytes
```

amiraabouhadid commented on packages/grid_client/tests/zos/deployments.test.ts:
```diff
@@ -0,0 +1,827 @@
+
+  CPU: 1,
+  MEMORY: 1024 * 1024 * 1024, // 1GB in bytes
+  SIZE: 10 * 1024 * 1024, // 10MB in bytes
```
```suggestion
  SIZE: 10 * 1024 ** 2, // 10MB in bytes
```

Commit:
Update packages/grid_client/tests/zos/deployments.test.ts

Co-authored-by: Amira <56790126+amiraabouhadid@users.noreply.github.com>

Commit:
Update packages/grid_client/tests/zos/deployments.test.ts

Co-authored-by: Amira <56790126+amiraabouhadid@users.noreply.github.com>

Commit:
- Add toByteArray function in helper
- Import mnemonic from the config file instead of dummy mnemonic
- Validate if mnemonic is not existed

samaradel commented on packages/grid_client/tests/zos/deployments.test.ts:
```diff
@@ -0,0 +1,827 @@
+      const deployment = createDeployment();
+
+      const challenge = deployment.challenge();
+      const expectedHash = md5(challenge).toString();
```
@AhmedHanafy725 agree ?

amiraabouhadid reviewed (approved):
<img width="262" height="66" alt="Screenshot 2025-07-13 at 11 50 51" src="https://github.com/user-attachments/assets/77b2c6ab-3c22-4f3b-aeae-dfeaa08d0964" />
